sdk/servicebus/azure-servicebus/azure/servicebus/aio/_async_auto_lock_renewer.py

```diff
@@ -0,0 +1,130 @@
+            return False
+        if not renewable._receiver._running:
+            return False
+        return True
+
+    async def _auto_lock_renew(self,
+                               renewable: "Union[ReceivedMessage, ServiceBusSession]",
+                               starttime: datetime.datetime,
+                               timeout: int,
+                               on_lock_renew_failure: "Optional[Callable[[Union[ServiceBusSession, ReceivedMessage], Optional[Exception]], Awaitable[None]]]"=None) -> None:
+        _log.debug("Running async lock auto-renew for %r seconds", timeout)
+        error = None
+        clean_shutdown = False # Only trigger the on_lock_renew_failure if halting was not expected (shutdown, etc)
+        try:
+            while self._renewable(renewable):
+                if (utc_now() - starttime) >= datetime.timedelta(seconds=timeout):
+                    _log.debug("Reached auto lock renew timeout - letting lock expire.")
+                    raise AutoLockRenewTimeout("Auto-renew period ({} seconds) elapsed.".format(timeout))
+                if (renewable.locked_until_utc - utc_now()) <= datetime.timedelta(seconds=self.renew_period):
+                    _log.debug("%r seconds or less until lock expires - auto renewing.", self.renew_period)
+                    await renewable.renew_lock()
+                await asyncio.sleep(self.sleep_time)
+            clean_shutdown = not renewable._lock_expired
+        except AutoLockRenewTimeout as e:
+            renewable.auto_renew_error = e
+            clean_shutdown = not renewable._lock_expired
+        except Exception as e:  # pylint: disable=broad-except
+            _log.debug("Failed to auto-renew lock: %r. Closing thread.", e)
+            error = AutoLockRenewFailed(
+                "Failed to auto-renew lock",
+                inner_exception=e)
+            renewable.auto_renew_error = error
+        finally:
+            if on_lock_renew_failure and not clean_shutdown:
+                await on_lock_renew_failure(renewable, error)
+
+    def register(self,
+                 renewable: "Union[ReceivedMessage, ServiceBusSession]",
+                 timeout: float = 300,
+                 on_lock_renew_failure: "Optional[Callable[[Union[ServiceBusSession, ReceivedMessage]], Awaitable[None]]]" = None) -> None:
+        """Register a renewable entity for automatic lock renewal.
+
+        :param renewable: A locked entity that needs to be renewed.
+        :type renewable: Union[~azure.servicebus.aio.ReceivedMessage,~azure.servicebus.aio.ServiceBusSession]
+        :param float timeout: A time in seconds that the lock should be maintained for.
+         Default value is 300 (5 minutes).
+        :param Optional[Callable[[Union[~azure.servicebus.aio.ServiceBusSession, ReceivedMessage]], Awaitable[None]]] on_lock_renew_failure: 
+         A callback may be specified to be called when the lock is lost on the renewable that is being registered.
```

Comment:
shall we mention only accepting async callback?

Reply:
So while in this case, Awaitable[None] is what indicates that fact (unless there's another notation I should be using?)

HOWEVER in double checking this, I noticed that I had used the async syntax even in the sync docs, so fixed that :)  (So nice catch!)